### PR TITLE
Add `reference.qubit` for testing and reference

### PR DESCRIPTION
**Pull request by astralcai**
Created from https://github.com/PennyLaneAI/pennylane/pull/5445

[sc-65558]

### CODE REVIEW

File: pennylane/devices/reference_qubit.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+        # Here we convert an arbitrary tape into one natively supported by the device
+        program = qml.transforms.core.TransformProgram()
+        program.add_transform(validate_device_wires, wires=self.wires, name="reference.qubit")
+        program.add_transform(qml.defer_measurements)
+        program.add_transform(qml.transforms.split_non_commuting)
+        program.add_transform(
+            decompose,
+            stopping_condition=supports_operation,
+            skip_initial_state_prep=False,
+            name="mini.qubit",
```
Comment (albi3ro) [inline pennylane/devices/reference_qubit.py:142]:
```suggestion
            name="reference.qubit",
```

**Review by albi3ro — COMMENTED** [review]
What would you think about adding some assertions inside `execute` to make sure we are passing in suitably supported tapes? For example:
```
assert tape.batch
assert all(supports_operation(op) for op in tape.operations)
```

Not mandatory, but might be useful.

File: pennylane/devices/reference_qubit.py [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+        program.add_transform(validate_device_wires, wires=self.wires, name="reference.qubit")
+        program.add_transform(qml.defer_measurements)
+        program.add_transform(qml.transforms.split_non_commuting)
+        program.add_transform(
+            decompose,
+            stopping_condition=supports_operation,
+            skip_initial_state_prep=False,
+            name="reference.qubit",
+        )
+        program.add_transform(qml.transforms.diagonalize_measurements)
```
Comment (albi3ro) [inline pennylane/devices/reference_qubit.py:144]:
```suggestion
        program.add_transform(qml.transforms.diagonalize_measurements)
        program.add_transform(
            decompose,
            stopping_condition=supports_operation,
            skip_initial_state_prep=False,
            name="reference.qubit",
        )
```

Should we switch the order?

File: tests/interfaces/test_jax_jit_qnode.py [hunk 3]
```diff
@@ -1508,6 +1513,8 @@ def test_hamiltonian_expansion_analytic(
         are non-commuting groups and the number of shots is None
         and the first and second order gradients are correctly evaluated"""
         gradient_kwargs = {}
+        if dev.name == "reference.qubit":
+            pytest.xfail("reference.qubit does not support hamiltonians")
```
Comment (albi3ro) [inline tests/interfaces/test_jax_jit_qnode.py:1517]: Isn't it calling `split_non_commuting` in preprocessing? Shouldn't that enable hamiltonian support?

Comment (albi3ro) [reply tests/interfaces/test_jax_jit_qnode.py:1517]: Is it just the `assert_no_called` part that fails, or something else?

Comment (astralcai) [reply tests/interfaces/test_jax_jit_qnode.py:1517]: The message was a bit misleading I guess, but the actual issue iirc was that the spying does not work on a transform. When `split_non_commuting` is replaced by the mocker, you can't add it to a transform program anymore, and preprocessing fails.

File: tests/interfaces/test_jax_qnode.py [hunk 4]
```diff
@@ -1445,6 +1452,13 @@ def test_hamiltonian_finite_shots(
         """Test that the Hamiltonian is correctly measured (and not expanded)
         if there are non-commuting groups and the number of shots is finite
         and the first and second order gradients are correctly evaluated"""
+
+        if dev.name == "reference.qubit":
+            pytest.skip(
+                "Cannot added to a transform to the transform program in "
```
Comment (albi3ro) [inline tests/interfaces/test_jax_qnode.py:1458]: And once again, not a huge fan of using mocker.spy for testing purposes...

**Review by albi3ro — APPROVED** [review]
❤️

**Review by lillian542 — COMMENTED** [review]
Should this be added to the `devices` so it turns up [in the documentation here](https://docs.pennylane.ai/en/stable/code/qml_devices.html)?

**Review by lillian542 — APPROVED** [review]
Looks great, thanks @astralcai! 🚀

**Pull request merged**